### ISSUE TITLE
PLEASE TEST: New Toolbars

### ISSUE THREAD

**Issue by IgnorantGuru** [IgnorantGuru/spacefm#519]
SpaceFM's new toolbars are ready for alpha testing in the new 'tool' branch [now merged into 'next'].  When you upgrade, this version will convert your old toolbar configuration, including any custom items, to the new style.  Please use a backup of your configuration for testing.

I would like to know how the upgrade converts your toolbar (details below), so I suggest running it like this:

~~~~
# make a safety backup of your config:
cp -a ~/.config/spacefm ~/.config/spacefm-1.0.1-backup

# make a backup of your config for testing:
rm -rf /tmp/spacefm-test
cp -a ~/.config/spacefm /tmp/spacefm-test

# start spacefm using the test config
spacefm -c /tmp/spacefm-test
~~~~

or, to test the default configuration, just start with a missing config dir:
~~~~
rm -rf /tmp/spacefm-new
spacefm -c /tmp/spacefm-new
~~~~

**To install the <del>'tool'</del> 'next' branch**, run these commands (you can always reinstall spacefm to downgrade):
~~~~
# If you'd like to test downgrades, save a copy of your current spacefm which you can run without rebuilding:
sudo cp /usr/bin/spacefm /usr/bin/spacefm101

# Download & Extract
mkdir /tmp/spacefm-next && cd /tmp/spacefm-next
wget -O spacefm-next.tar.gz https://github.com/IgnorantGuru/spacefm/tarball/next
tar xzf spacefm-next.tar.gz
cd IgnorantGuru-spacefm-*

# Build & Install
./configure --prefix=/usr  # and optionally: --with-gtk3 see ./configure --help
make -s
sudo make install
sudo update-mime-database /usr/share/mime > /dev/null
sudo update-desktop-database -q
sudo gtk-update-icon-cache -q -t -f /usr/share/icons/hicolor
sudo gtk-update-icon-cache -q -t -f /usr/share/icons/Faenza

# Remove Temporary Files
cd / && rm -rf /tmp/spacefm-next
~~~~

You can copy and paste all of those commands into your terminal as a set. Because this branch is still under some development, it is best to update it every few days by repeating the above commands, and check <a href="https://github.com/IgnorantGuru/spacefm/milestones/1.0.2">issues related to the tool branch</a> and <a href="https://github.com/IgnorantGuru/spacefm/commits/next">see latest changes</a>.

Or, if you are familiar with git, you can clone the whole SpaceFM repo at git@github.com:IgnorantGuru/spacefm.git **or** https://github.com/IgnorantGuru/spacefm.git

**What's New**

<ul>
    <li>Design Mode can now be used by right-clicking on toolbar items.  The toolbar configuration menus have been removed.<br><br>
    <li>Builtin toolbar items can be reordered.<br><br>
    <li>Items may be cut/copied/pasted to/from the booksmarks pane, toolbars, and menus.<br><br>
    <li>Each panel now has its own independent toolbars, although they should all look the same at the start.<br><br>
    <li>Tooltips are now optionally shown for toolbar items.<br><br>
    <li>A new Show Hidden builtin toolbar item is available to toggle showing hidden files.<br><br>
    <li>Note that the keyboard can no longer be used on the toolbar (eg focus and select).  Not sure if this is needed but I can add a menu, perhaps in the context menu of the path bar, which would mirror the toolbars and allow you to change them via keys only (sort of like the old toolbar config menus).  Let me know if anyone wants this, it's pretty simple to add.<br><br>
    <li>When you upgrade to this version, your old toolbars will be converted to the new config.  Any custom items you added to the toolbar will be included, and will be shown whether Show was enabled for them or not.  (The new toolbars have no Show setting.)  Also, any custom items you added in the toolbar config menus will be moved to a Lost+Found submenu in the main Tools menu.<br><br>
    <li>When downgrading, any custom toolbar items will vanish from view.  When you upgrade again, they should reappear.<br><br>
    <li>The manual has been updated, especially the <a href="http://ignorantguru.github.io/spacefm/spacefm-manual-next-en.html#designmode-toolbars">Toolbars</a> section.<br><br>
</ul>

Thanks for testing!

**Comment by IgnorantGuru** [comment]
Changes in the tool branch have been merged into next, and the tool branch will be deleted shortly.  For install instructions for the next branch, please see BUILD NEXT in README.

**Comment by arclance** [comment]
Padding on the toolbar icons using GTK2 seems to have changed (increased).
It might be using a different part of my GTK2 theme I never saw a program use before so I never customized the padding in that area though.
I will check that now.

**Comment by IgnorantGuru** [comment]
@arclance I'm not seeing any increase in toolbar padding, GTK2 or GTK3, default or 3rd party themes.

The way the toolbar is built has changed.  They are now a generic GtkButton packed into a GtkEventBox packed into a GtkToolItem, rather than merely a GtkToolButton.

**Comment by arclance** [comment]
Maybe the padding is adding together because of the nesting?

**Comment by IgnorantGuru** [comment]
It shouldn't.  The GtkEventBox and GtkToolItem are just object containers that don't have any visual aspect.  The button should be the only thing receiving any padding.

I did notice in my 3rd party GTK2 theme that the background color behind the button is now a slightly different shade than before.  I don't see any appreciable change in GTK2 themeless.

**Comment by IgnorantGuru** [comment]
I think [this event box setting](https://developer.gnome.org/gtk3/stable/GtkEventBox.html#gtk-event-box-set-visible-window) may be responsible for the toolbar color change, and maybe the padding in some themes.  I'll turn visibility off - it's an odd default.

**Comment by IgnorantGuru** [comment]
Event boxes are now not visible - this removed the shading I was seeing behind the toolbar.

**Comment by arclance** [comment]
That looks much better.
The buttons are much more obviously separate now.

For reference here is a screenshot of the toolbar from a late 0.9.x next branch build.
![toolbar_icons_before_tool](https://cloud.githubusercontent.com/assets/1261154/7661353/d61f6012-fb1f-11e4-941e-dd4beabc3119.png)
Here is a screenshot of the next branch just after the tool branch was merged into it.
![toolbar_icons_after_tool](https://cloud.githubusercontent.com/assets/1261154/7661362/e75b4152-fb1f-11e4-9718-b7c352fe5c7c.png)
And here is a screenshot after commit 80dcd09.
![toolbar_icons_after_tool_change_ after_commit_80dcd09](https://cloud.githubusercontent.com/assets/1261154/7661364/ee53c4f2-fb1f-11e4-8f48-5ac346755d28.png)

The current buttons are larger than before but that is probably because my GTK2 theme has different padding values for GtkButton and GtkToolButton.

**Comment by IgnorantGuru** [comment]
Thanks for the screenshots.  i don't see any other padding that can be reduced.  And this method with the buttons is the only one where I've been able to get a right-click event.  GTK3/Adwaita is where the padding seems a bit more extreme, but GTK3 is like that in general.

**Comment by IgnorantGuru** [comment]
Actually I did find a way to crunch the GTK2 toolbar down using a size request on the buttons, to the point where the icons are on top of each other.  In  fb6f37c I have padding set to 4 for toolbar icon size 24x24 and larger, and 6 for < 24x24.  This was needed to prevent clipping the icons (the button needs to be slightly larger than the icon).  I'm a little unsure how hardcoding these values will affect themes and icons, so any testing will be helpful.  You can also try adjusting the padding values [here](https://github.com/IgnorantGuru/spacefm/blob/fb6f37c0740f393bda60a8deacfb1e2bbe9da831/src/settings.c#L9658).  Different padding values are used for GTK2 vs GTK3 there.

GTK3 didn't respond at all, and I also set all margins to 0 (a GTK3-only feature which apparently does nothing), and I set the toolbar container to have zero border (GTK2 and GTK3).  The only thing which had an effect on GTK3 was setting the padding (size request) higher than zero, which added more.  The Adwaita theme is definitely more wasteful of space than other GTK3 themes I tried (wastes about 3mm vertical compared to thinnest theme I tried).

**Comment by IgnorantGuru** [comment]
Note, I have reverted the size requests (manual padding) on buttons because it was causing clipping of icons in GTK2, eg with no theme or other themes.  It's not good to hard-code such things because they are theme-dependent, so I think leaving it as-is is the best approach.  The code is still in there, I just commented it out for now.  And I left the GTK3 margin settings and toolbar border setting intact, in case they do some good.

If you would like to test the build that had the size requests and padding I mentioned above, you can download that branch like this:

     wget -O spacefm.tar.gz https://github.com/IgnorantGuru/spacefm/tarball/fb6f37c0

Then do the normal configure and make.

**Comment by IgnorantGuru** [comment]
Actually a padding of 10 (h and v) for all toolbar sizes seems to be a reasonable medium, and it does compress the toolbar nicely.  I have done this for GTK2 only in 87791bb.  GTK3 doesn't seem to respond so I prefer not to set it, in case they wildly change the behavior in a future version, but if you would like to test it with GTK3 just use fb6f37c.  

My only (GTK2) question is, if a theme has high button padding, will the icons get clipped?  If @arclance or someone else can answer that question I'd feel better.  Be sure to try all the toolbar sizes.

**Comment by IgnorantGuru** [comment]
Okay I pushed a number of commits there, mainly because I kept changing my mind, but you can also use those to test the variations.  As I explained above, you can build the branch at any commit - just substitute the commit number for 'next' wen downloading the tarball.

The way I left it is that it leaves the button height natural (size request -1), and sets the width to the icon size plus a padding value.  Leaving the height natural prevents clipping of icons that was still occuring at some sizes.  I still don't know what will happen if a theme has a lot of button padding.

All of that is GTK2 only - GTK3 still has the size request removed.

**Comment by IgnorantGuru** [comment]
For **GTK3**, I did figure out how to remove the padding from the toolbar and buttons in 84b38e9.  This works for me with Adwaita - actually the toolbar comes out shorter than it was in 1.0.1 before the toolbars were changed.

I did add 1 px of padding to the top and bottom of the toolbar, otherwise the toggle buttons merged with the tab bar.

Also, the horizontal padding was removed from the buttons, which means more of them fit in the toolbar.

**Comment by HerborenInVlammen** [comment]
Hi again:  I have spent the last few hours truly messing with the new toolbars, and I am finding nothing to complain about... so something is definitely wrong here (joke).  Very, very cool actually.

Anyway, just a quick question:  Is there any way to add a built-in command to the toolbar?  I am having no luck.  What I want it to be able to add "New Archive" and "Extract To" to the toolbars.  I have tried creating the commands from scratch but it is not working or I am doing it wrong.

**Comment by IgnorantGuru** [comment]
@HerborenInVlammen Thanks for testing.  There is no provided way to add built-in menu functions to any menu or toolbar.  Only built-in toolbar items may be added to toolbars.

You can do this by setting a shortcut key to the menu item you want.  Then create a custom command which emits the shortcut key of that item, using socket command [emit-key](http://ignorantguru.github.io/spacefm/spacefm-manual-en.html#sockets-methods-emit-key).

**Comment by IgnorantGuru** [comment]
**Additions**:  Two new built-in toolbar items:  Show Thumbnails and Large Icons.  Also, Show Thumbnails has been added to the panel's View|Style menu, and the panel's View menu has been added to the main View menu (new users have a hard time finding it in the context menu).

@HerborenInVlammen Also, many menu items can be duplicated as custom commands using socket commands (examples in the manual).

**Comment by HerborenInVlammen** [comment]
Thank you again.  It works perfectly :)

**Comment by arclance** [comment]
What you did knocked GTK2 toolbar icons off center.
![space_fm_tool_gtk2_update](https://cloud.githubusercontent.com/assets/1261154/7703163/f05e7102-fe03-11e4-9e34-ec5b40f6087c.png)
Looks kind of like it only affected the right side of the button.

**Comment by HerborenInVlammen** [comment]
Well, it was not intended for GTK2.  Only GTK3 as per the post.  :)

**Comment by arclance** [comment]
@HerborenInVlammen and @IgnorantGuru 
I was referring to this which is for GTK2 not GTK3.

```
Okay I pushed a number of commits there, mainly because I kept changing my mind, but you can also use those to test the variations. As I explained above, you can build the branch at any commit - just substitute the commit number for 'next' wen downloading the tarball.

The way I left it is that it leaves the button height natural (size request -1), and sets the width to the icon size plus a padding value. Leaving the height natural prevents clipping of icons that was still occuring at some sizes. I still don't know what will happen if a theme has a lot of button padding.

All of that is GTK2 only - GTK3 still has the size request removed.
```

 I did not have a chance to test it until this morning since we don't use Linux at work.

**Comment by Vladimir-csp** [comment]
At least with GTK2 the new toolbars now lack button interactivity (hover, press effects)

**Comment by IgnorantGuru** [comment]
Jfyi, the toolbars and bookmarks should now work with icon set to an image file path (eg from the new icon chooser), should honor the 'icon' file of custom items, and menu item icons from a file should now be scaled.  This should create greater consistency in how icons are shown in various places.

@arclance 
> What you did knocked GTK2 toolbar icons off center.

Indeed, that competes with minimizing the padding.  I have adjusted it slightly upward to compensate.

@Vladimir-csp 
> At least with GTK2 the new toolbars now lack button interactivity (hover, press effects)

Yes, this is so in both GTK2 and GTK3, in addition to the loss of the ability to navigate the toolbar by keyboard, and the overflow menus.

**Comment by arclance** [comment]
Indeed, that competes with minimizing the padding. I have adjusted it slightly upward to compensate.

I updated today and don't see any change to the position of the toolbar button icons in GTK2.
They are still off center, it does not look like the change had any effect.

**Comment by IgnorantGuru** [comment]
@arclance 
> I updated today and don't see any change to the position of the toolbar button icons in GTK2.
They are still off center, it does not look like the change had any effect.

You can adjust the padding values [here](https://github.com/IgnorantGuru/spacefm/blob/47f41719f69c9f119bc6de98609a64ae3e3c4a3a/src/settings.c#L9754) if you'd like to see what works.  I think if we add much more padding then we may as well just use the default theme padding and not hack it.  It may also be possible to adjust the image alignment, or remove the padding that the theme is adding.  Perhaps your buttons have a lot of padding which is shifting it when the button is sized smaller.  They aren't off-center for me anymore.

**Comment by arclance** [comment]
You can adjust the padding values here if you'd like to see what works. 
    I think if we add much more padding then we may as well just use the default theme padding and not hack it.

Looking at this

     if ( real_icon_size < 18 )
    hpad = 9;

I wonder if the fact that my theme uses 12x12 toobar icons might not work correctly with a hpad value of 9 (75% of the icon size) and is just pushing the icons to the right side of the button and GTK2 won't allow it to be pushed any farther.
I had to work late today so I don't have time to test what happens if I try hardcoding smaller hpad values myself tonight.

    Perhaps your buttons have a lot of padding which is shifting it when the button is sized smaller.

Padding for things like this in my theme varies from 0 to 3 pixels for comparison.
From what you have said in the past about how you made the buttons with GTK2 I think the padding should be set to 3 in my theme in this case.

**Comment by arclance** [comment]
@IgnorantGuru 
Any updates?
I could not get the icons back on center so this might be the way to go.

    I think if we add much more padding then we may as well just use the default theme padding and not hack it.

**Comment by Vladimir-csp** [comment]
>>    At least with GTK2 the new toolbars now lack button interactivity (hover, press effects)
>
>Yes, this is so in both GTK2 and GTK3, in addition to the loss of the ability to navigate the toolbar by >keyboard, and the overflow menus.

Is it permanent and inevitable?

**Comment by IgnorantGuru** [comment]
@arclance I did make an adjustment to this in next.  If that's still a problem then it may need to be removed.  This is what usually happens when you hack styles - it ends up working for one theme but not another.  I'll take another look at it - might try an alignment adjustment.

@Vladimir-csp Effects were lost because of the method I used to get a right-click event on a toolbar button.  There are other more involved ways of doing this, such as creating a custom widget, but at least for now that's how it's implemented. The GTK API is not very helpful in this area (as if they can't imagine anyone right-clicking on a button). The buttons are now just visual props that lose their built-in methods because the mouse events (including hover) are intercepted before reaching them.

Overall, this has been a problem area in development of the toolbars.  When originally developing SpaceFM, I eventually gave up on it and moved on, using the old toolbar config menus instead of a right-click.  So I would say this is an improvement if not perfect.  Should probably be done with custom widgets, but that's a bit more involved.

**Comment by IgnorantGuru** [comment]
@arclance The other way I could approach it is to remove the theme's button padding then add our own, like I did for GTK3, but using GTK2's style methods.  This is more likely to succeed than just forcing the size of the button like I'm doing now.  Might give that a try.

**Comment by arclance** [comment]
@IgnorantGuru 
If you mean https://github.com/IgnorantGuru/spacefm/commit/360e439dae20f907b9a313a014cc522443c963d3 or https://github.com/IgnorantGuru/spacefm/commit/47f41719f69c9f119bc6de98609a64ae3e3c4a3a they did nothing with my theme.

**Comment by IgnorantGuru** [comment]
In bba34ed, I merely trim 4 pixels from the GTK2 button widths, only if the buttons are not checkbox style.  Please see if that corrects the theme issues.

**Comment by arclance** [comment]
Yes the icons are back in the right place.
The buttons look basically the same as they did with commit https://github.com/IgnorantGuru/spacefm/commit/80dcd09dc5de0c44a233adfa95359cc487241b90.

**Comment by IgnorantGuru** [comment]
Thanks for all the testing - this has been released in 1.0.2.
